File: 1 - Introduction & stratégies de recherche/Problèmes/H - Fair Division/A.c
#include<stdio.h>
#define SIZE 100000

int subset_sum(short B[], int n, int sum) {

	short table[n + 1][sum + 1]; 
	
	int i;
	int j;
	// si la somme vaut 0 alors vrai 
	for (i = 0; i <= n; i++) 
		table[i][0] = 1; 

	// si la somme attendue n'est pas 0 mais que le set est vide : false
	for (i = 1; i <= sum; i++) 
		table[0][i] = 0; 

	// DP : remplir la table
	for (i = 1; i <= n; i++) { 
		for (j = 1; j <= sum; j++) { 
			if (j < B[i - 1]) 
				table[i][j] = table[i - 1][j]; 
			if (j >= B[i - 1]) 
				table[i][j] = table[i - 1][j] || 
							table[i - 1][j - B[i - 1]]; 
		} 
	}
	return table[n][sum]; 
} 

int main(void){

	int t;
	scanf("%d", &t);
	
	short B[SIZE];
	int i;
	int n;
	int j;
	int s;
	for(i=0; i < t; ++i) {
		s=0;
		scanf("%d", &n);
		
		for(j=0; j < n; ++j) {
			scanf("%hu", &B[j]);
			s+=B[j];
		}
		
		if (((s % 2)==0) && subset_sum(B, n, s/2))
			printf("YES\n");
		else
			printf("NO\n");
	}

	return 0;

}

File: 1 - Introduction & stratégies de recherche/Problèmes/H - Fair Division/A.py
def subset_sum(b, n, s):
	table = [[False]*(s+1)]*(n+1)
	
	# si la somme vaut 0 alors vrai 
	for i in range(n+1):
		table[i][0] = True
	# si la somme attendue n'est pas 0 mais que le set est vide : false
	for i in range(1, s+1):
		table[0][i] = False

	# DP : remplir la table
	for i in range(1, n+1):
		for j in range(1, s+1):
			if j < b[i-1]:
				table[i][j] = table[i-1][j]
			if j >= b[i-1]:
				table[i][j] = (table[i-1][j] or table[i-1][j-b[i-1]])

	return table[n][s]
	
	

t = int(input())

for i in range(t):
	
	s=0
	n = int(input())
	
	b = list(map(int, input().split()))
	s = sum(b)
	if s%2==0 and subset_sum(b, n, int(s/2)):
		print("YES")
	else:
		print("NO")
